Docs & Feedback Update (#10)
adobe-rnd/aem-gtm-martech: README.md

  dataLayerInstanceName: /* Name of the DataLayer to use for all events. (default 'gtmDataLayer') */,
  tags: [/* One or more GA4 Measumrent Ids */],
  containers: {
    lazy: [/* Zero or more GTM Container Ids to load during Lazy Phase */],
    lazy: [/* Zero or more GTM Container Ids to load during Delayed Phase */],
  },
  pageMetadata: { /* Metadata to pass on during the intializaton of the GA4 tag */ },
  consent: /* require consent (default: enabled) */,
  consentCallback: /* Function that handles consent processing, if consent is enabled, this must be specified */,
  decorateCallback: /* Function to call on each found or loaded Section/Block */,
});

```
Note that:
- If `consent` is enabled, then by default all [consent types](https://developers.google.com/tag-platform/security/concepts/consent-mode#consent-types) will be set to `denied`. 
- If `consent` is enabled, then a `consentCallback` must be specified. 

### 4. Call Eager Phase Function

Update the `loadEager` function in your `scripts.js` file, to call plugin's eager phase:

```js
async function loadEager(doc) {
  …
  if (main) {
    decorateMain(main);

>     await eager();
>     doc.body.classList.add('appear');
>     await loadSection(main.querySelector('.section'), waitForFirstImage);

Reviewer: I'd be tempted to do:
```suggestion
    doc.body.classList.add('appear');
    await Promise.all([
      eager(),
      loadSection(main.querySelector('.section'), waitForFirstImage),
    ]);
```

that way we load the gtm in parallel while also loading the LCP, instead of being blocking

  }
  …
}
```

> Note that the `eager()` call is asynchronous, therefore can be added before or after the LCP section. We do recommend awaiting it, as future updaets (such as personalization support) may require it.

Reviewer: ```suggestion
Note that the `eager()` call is asynchronous, therefore can be added before or after the LCP section. We do recommend awaiting it, as future updates (such as personalization support) may require it.
```

  - `pageMetadata` `{Object}`: A set of key-value pairs to pass to the GA4 tag initializer.
  - `consent` `{Boolean}`: Enable consent. Default: `true`
  - `consentCallback` `{Function|undefined}`: A function that will perform consent validation. Returns a promise that resolves to an object, which will be passed to the GA for update.
  - `decorateCallback` `{Function}`: A function that can decorate HTML elements for events. Passed all sections & blocks found.

---

### `pushToDataLayer(payload)`
Pushes a generic payload to the Adobe Client Data Layer.

- **`payload`** `{Object}`: The data object to push.

---

### `updateUserConsent(consent)`
Updates the consent according to the []`gtag.js` implementation](https://developers.google.com/tag-platform/security/guides/consent?consentmode=advanced#implementation_example)

- **`consent`** `{Object}`: An object detailing user consent choices.

---

## An Example Site

An example of this plugin in use can be found on the [AEM GTM Martech demo site](https://main--aem-gtm-martech-site--adobe-rnd.aem.page/).

## FAQ

### I need the page view data immediately, when is it captured?

> All GA4 scripts are imported to the page during the Eager Phase. During this process a page view event is dispatched, along with any custom metadata provided to during plugin initialzation. Even if the GTM containers are loaded lazily, this data will be captured as soon as Google's library is conifgured to dispatch it.

Reviewer: It will only be submitted when the GTM loads though, which means the page view is reported at the end of the lazy as a "bottom of page event" one the page is fully rendered. This is also what we do for the Adobe stack

Author: I tested this and didnt load any GTM libraries and it still sends the page view. According to docs, that is supposed to happen using the gtag function.

Reviewer: ok, nice! Conceptually, I'm still undecided if this is right though. The page is not rendered until end of "lazy"… so technically it was not viewed in "eager" yet. Might be nitpicking, but to me that's kinda misleading 😅

Author: Hm. So technically the view event isn't dispatched until GA decides to send the first batch of events. So maybe:

```suggestion
All GA4 scripts are imported to the page during the Eager Phase. During this process a `page_view` event occurs, which includes any custom metadata provided to during plugin initialization. This `page_view` event will be dispatched, as soon as the Google library decides it should be sent, regardless of when any other phase's GTM libraries are loaded.
```

Thoughts?

Reviewer: Approved :)
